Use backwards compatible term_to_binary

martinsumner opened the pull request:
So that where we have hashed term_to_binary output in OTP25 or earlier, that has will be matched in OTP 26.

martinsumner committed:
Use backwards compatible term_to_binary

So that where we have hashed term_to_binary output in OTP25 or earlier, that has will be matched in OTP 26.

ThomasArts reviewed (commented):
It is hard to see if you covered all cases.

I did a grep on term_to_binary to see if there might be places that you could have forgotten... One suspect:

[src/leveled_inker.erl:    ManifestBin = term_to_binary(Manifest),
A manifest has a key as last element of the tuple:](https://github.com/martinsumner/leveled/blob/bc87273c76cd7fb86c40a528cbcc1a8cd8d9eb61/src/leveled_inker.erl#L1331C1-L1335C44)
According to the code in leveled_cdb:
https://github.com/martinsumner/leveled/blob/bc87273c76cd7fb86c40a528cbcc1a8cd8d9eb61/src/leveled_cdb.erl#L824C1-L825C66
the last key may have value `empty`. 

So this works fine when you write to disk in OTP25 and then open in OTP26, reading with binary_to_term... however, if OTP26 writes it, you won't be able to open it again with the OTP25 version.

So, there might be some smaller issues left after this refactoring.

martinsumner commented:
I'm not sure that is a problem.  As I understand it the binary_to_term/term_to_binary is still backwards/forwards compatible between OTP 25 and OTP 26 (as OTP 25 already supported the unicode encoding of atoms, it just wasn't the default behaviour).  The only concern is where we:

- Use term_to_binary;
- Hash the result of using term_to_binary;
- Persist the result hash as part of file structure (e.g. in a bloom filter) so that we expect the hash to match any future request.

In this case if we save the file in OTP 25, and then open when we're running OTP 26, it will trust the bloom filter that it reads off disk.  If OTP 26 then hashes the same key (after converting to binary) and we check for presence of the key in the bloom filter, if OTP 26 hash has a subtle change in binary_to_term, then the hash will be different, and the code will assume the key was not present (due to bloom filter mismatch) when actually it is.

So as long as there is always a hash involved we should be OK, as long as too big a jump in OTP version.  I think the new encoding in OTP 26 can be read using term_to_binary as far back as OTP R16B.

martinsumner committed:
Merge branch 'develop-3.1' into mas-i407-fixt2b

martinsumner committed:
Test reliability

If all keys are put in order, the max_slots may not be used, as the driver at L0 is penciller cache size, and merge to new files (managed by the parameter) only occurs when there are overlapping files the level below
